### ISSUE TITLE
When defining subclass of Array, unrelated code stops compiling

### ISSUE THREAD

**Issue by Fryguy** [crystal-lang/crystal#7447]
```
Crystal 0.27.2 (2019-02-05)

LLVM: 6.0.1
Default target: x86_64-apple-macosx
```

The following code is the smallest demonstration I could come up with (also at https://play.crystal-lang.org/#/r/6a03).  It fails to compile as is with the error further down, but if you comment out the little class definition, then it complies just fine.  Note that my original code, after calling yield, would set values into the `@matrix`, but as I was whittling this away, I realized it wasn't needed to demonstrate the issue...only mentioning this in case there's something else going on where that is important.

```crystal
# Comment this to hide the error
class Foos < Array(String)
end

class Matrix
  def initialize
    @matrix = Array(Array(Float64)).new(2) { Array(Float64).new(2, Random.rand) }
    yield
  end

  def transpose
    Matrix.new { 2.times.map { |x| @matrix.map &.[x] } }
  end
end

m = Matrix.new { }
m.transpose
```

```text
Error in error.cr:16: instantiating 'Matrix#transpose()'

m.transpose
  ^~~~~~~~~

in error.cr:11: instantiating 'Matrix.class#new()'

    Matrix.new { 2.times.map { |x| @matrix.map &.[x] } }
           ^~~

in error.cr:11: instantiating 'Matrix.class#new()'

    Matrix.new { 2.times.map { |x| @matrix.map &.[x] } }
           ^~~

in error.cr:11: instantiating 'Int::TimesIterator(Int32)#map()'

    Matrix.new { 2.times.map { |x| @matrix.map &.[x] } }
                         ^~~

in /usr/local/Cellar/crystal/0.27.2/src/iterator.cr:710: instantiating 'Iterator::Map(Int::TimesIterator(Int32), Int32, Array(Float64)).class#new(Int::TimesIterator(Int32), Proc(Int32, Array(Float64)))'

    Map(typeof(self), T, U).new(self, func)
                            ^~~

in /usr/local/Cellar/crystal/0.27.2/src/iterator.cr:717: instance variable '@func' of Iterator::Map(Int::TimesIterator(Int32), Int32, Array(Float64)) must be Proc(Int32, Array(Float64)), not Proc(Int32, Array(Float64))

    def initialize(@iterator : I, @func : T -> U)
                                  ^~~~~
```

---

I *think* this happens because `2.times.map { ... }` seems to return an Iterator, even though I would expect map to resolve to an Array(TheBlocksType) when given a block.  However, what I can't understand is how it compiles before the unrelated class definition but not after.  https://play.crystal-lang.org/#/r/6a04

```crystal
x = [[1.0, 1.1], [2.0, 2.1]]
puts 2.times.map { |n| x.map &.[n] }
puts 2.times.map { |n| x.map &.[n] }.class

puts 2.times.map { |n| x.map &.[n] }.to_a
puts 2.times.map { |n| x.map &.[n] }.to_a.class

puts 2.times.map { |n| n * 2 }
puts 2.times.map { |n| n * 2 }.to_a
```

```text
Iterator::Map(Int::TimesIterator(Int32), Int32, Array(Float64))(@iterator=#<Int::TimesIterator(Int32):0x55dc589e7fd0 @n=2, @index=0>, @func=#<Proc(Int32, Array(Float64)):0x55dc57ef5400:closure>)
Iterator::Map(Int::TimesIterator(Int32), Int32, Array(Float64))
[[1.0, 2.0], [1.1, 2.1]]
Array(Array(Float64))
Iterator::Map(Int::TimesIterator(Int32), Int32, Int32)(@iterator=#<Int::TimesIterator(Int32):0x55dc589e7f90 @n=2, @index=0>, @func=#<Proc(Int32, Int32):0x55dc57ef59f0>)
[0, 2]
```

Might be related to #6789

**Comment by asterite** [comment]
This has to do with the several bugs related to generic types and inheritance. For now my advise is: don't do it.

**Comment by Sija** [comment]
How about using some money from the donations to setup bounties for the most longstanding bugs (like the ones related to generic types) we have?

**Comment by Fryguy** [comment]
I'm not really held up on it, but it threw me for a bit because of the unrelated nature of it, so I wanted to make sure I opened an issue for it.

I found that if I just defined my `Foos` class as:

```crystal
class Foos
  @internal = Array(String).new

  delegate :<<, :[], size, to: @internal
end
```

That worked for my use cases, though it's definitely not as elegant.

**Comment by Sija** [comment]
@Fryguy inheriting from stdlib container classes already proved many times to be the source of weird compiler errors, so at least for the time being it's a big no-no and IMHO should be noted in some visible place prepended with big a ⚠️ sign...

**Comment by straight-shoota** [comment]
> How about using some money from the donations to setup bounties for the most longstanding bugs (like the ones related to generic types) we have?

I'm not sure if that would be generally feasible, but in this case it probably wouldn't necessarily help because the entire generics system needs to be revisited (and probably also re-implemented) anyway. And that's not a task you could set a bounty for.

**Comment by washu** [comment]
use 

class Foos(T) < Array(T)
end

**Comment by asterite** [comment]
Fixed by #8159
Duplicate of #6789

**Comment by RX14** [comment]
Perhaps we should have some kind of sealed attribute which we can place on stdlib classes which are not meant to be inherited.

We can't ensure code which inherits *many* stdlib classes will keep compiling after `1.0`, so why not make that formal?

**Comment by crysbot** [comment]
This issue has been mentioned on **Crystal Forum**. There might be relevant details there:

https://forum.crystal-lang.org/t/why-is-this-failing/7161/2
